### ISSUE TITLE
Different results since v2

### ISSUE THREAD

**Issue by hlesemann** [libscran/umappp#24]
Since upgrading to version 2 I seem to be getting rather strange results with the same data.
![image (1)](https://github.com/user-attachments/assets/795a418f-a447-49fd-a7a6-e98363b2fdcc)
This is what it looked like and what I would expect from umap.
In contrast this what I get now with the same data.
![image](https://github.com/user-attachments/assets/ad59b5aa-0eca-4518-8884-77a8cc794390)
Any idea what could cause a structure like this?

**Comment by LTLA** [comment]
Hm. Nothing obvious comes to mind. The only intended change was an increase in `min_dist` to match the defaults from **uwot**. 

If I had to speculate, I would say that this looks a bit like the UMAP immediately after spectral initialization, i.e., no epochs. Did you call `run()` on the `Status` object?

**Comment by hlesemann** [comment]
Yes sure. I was trying around with the options and was able to get better results, but they always looked kind of rectangular in a way.
<img width="1189" alt="Screenshot 2024-11-13 at 18 07 01" src="https://github.com/user-attachments/assets/5bbe2874-7285-4410-9401-72434eb70622">

**Comment by LTLA** [comment]
I have no idea - I don't place any bounds on the coordinates.

Looking through the code, the only other intended change was how the a/b values were calculated, but that should just involve a change in the algorithm to improve convergence.

If you can send me a MRE, I can have a look at it. Just the nearest neighbor results (i.e., indices and distances) should be sufficient.

**Comment by jlmelville** [comment]
If you can plot the actual value of the coordinate axes that can help diagnose issues: an initialization where the interpoint distances are large could lead to problems where most edges get effectively zero gradient and never update.

Points coalescing at the edges of a plot can also be a sign of a too-high learning rate although I am surprised that the UMAP gradient does this as I have usually only experienced this with different dimensionality embedding methods (idiotic ones of my own invention usually).

**Comment by hlesemann** [comment]
Yes this is what I got with num_epochs set to zero:
<img width="1208" alt="Screenshot 2024-11-14 at 08 53 42" src="https://github.com/user-attachments/assets/452d7d02-67ff-4889-8029-a4324a3852c3">
Can you give an example what you mean by giving you the nearest neighbors? 
Don't know if this also helps but here's an extract of the dataset
[mre.json](https://github.com/user-attachments/files/17745375/mre.json)

**Comment by LTLA** [comment]
I'm assuming that, in your `mre.json`, each inner array contains the coordinates for an observation... but it seems that different observations have different numbers of coordinates. In R:

```r
out <- jsonlite::fromJSON("mre.json", simplifyVector=FALSE)
table(lengths(out))
##   13   20
##  689 4263
```

In other words, most observations have 20 coordinates but others have 13. I'm not sure how you're doing the neighbor detection across different numbers of coordinates, but **umappp** won't do anything for missing values.

Nonetheless, I tried to proceed by just taking the first 13 coordinates for each observation and running it through. (For convenience, I'm using the **umappp** R package in [`tests/R/package`](https://github.com/libscran/umappp/tree/master/tests/R/package), which just wraps the C++ code.)

```r
mat <- do.call(rbind, lapply(out, function(x) head(unlist(x), 13)))
nn <- BiocNeighbors::findKNN(mat, k=15)
coords <- umappp::runUmap(nn$index, nn$distance)
plot(coords[,1], coords[,2])
```
![test](https://github.com/user-attachments/assets/e0ac3ce5-935d-4843-a448-46979da6c168)

Looks pretty UMAP-ish to me.

**Comment by hlesemann** [comment]
Okay the dataset I sent was wrong, I am sry for that. My algorithm wasn't affected by this because it eliminated the other values though. 


```
std::vector<std::vector<double>> UMAP::runEmbedding(std::vector<std::vector<double>> featureMatrix)
{
    std::vector<double> featureColumnMajor;
    
    std::vector<std::vector<double>> coordinates;
    int rows;
    int columns;
    rows = featureMatrix.size(); // amount of entries in data
    columns = featureMatrix[0].size(); //values for each entry
    for (int j = 0; j < columns; j++)
    {
        for (int i = 0; i < rows; i++)
        {
            featureColumnMajor.push_back(featureMatrix[i][j]);
        }
    }
    size_t out_dim = 2;
    std::vector<double> embedding(rows * 2);
    umappp::Options opt;


    auto status = umappp::initialize(
        columns,
        rows,
        featureColumnMajor.data(),
        knncolle::VptreeBuilder(), 
        out_dim,
        embedding.data(),
        opt
    );
    status.run();

    //creating 2 dimensional vector (coordinates)
    for (int i = 0; i < rows; ++i)
    {
        std::vector<double> coordinate;
        coordinate.push_back(embedding[i]);    
        coordinate.push_back(embedding[rows+i]); 
        coordinates.push_back(point);           
    }
    return coordinates;
}
```

Anything obvious that comes to your minds? I was thinking about the final formatting back to a 2d array e.g.

**Comment by LTLA** [comment]
The `initialize()` overload that you're using expects the number of dimensions as the first argument, then the number of observations, for a column-major matrix. It seems that you have `rows` and `columns` mixed up here, as the fastest-changing dimension (i.e., the rows) should correspond to the inner loop over `i`.

The final embedding is also column-major where rows are the dimensions and columns are the observations; so `embedding[2*j]` and `embedding[2*j + 1]` would be the coordinates for observation `j`.

**Comment by hlesemann** [comment]
Damn I already tried both separately.. Thanks a lot, that resolves my problem

**Comment by hlesemann** [comment]
Issue was in implementation on users end
